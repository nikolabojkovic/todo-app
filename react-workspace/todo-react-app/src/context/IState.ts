import { ITodo } from "../models/Todo";
import { IPaging } from "../models/IPaging";
import { IFilter } from "../models/IFilter";
import { ISort, SortDirection } from "../models/ISort";
import { ISearch } from "../models/ISearch";
import { StateFilter } from '../models/IFilter';
import { IAction } from "../models/Action";
import { ISettings, IGeneralSettings, IPaginationSettings, ISearchSettings, PaginationType, ListContainerType } from "../models/ISettings";

export enum DisplayMode {
	All,
	Filtered
}

export interface IState {
	isLoading: boolean;
	originalList: ITodo[];
	displayList: ITodo[];
	displayMode: DisplayMode;
	effectTrigger: IAction | null;
	search: ISearch;
	filter: IFilter;
	sort: ISort;
	paging: IPaging;
	settings: ISettings;
	activeTab: string;
}

export class State implements IState {
	isLoading: boolean;
	originalList: ITodo[];
	displayList: ITodo[];
	displayMode: DisplayMode;
	effectTrigger: IAction | null;
	search: ISearch;
	filter: IFilter;
	sort: ISort;
	paging: IPaging;
	settings: ISettings;
	activeTab: string;

	constructor(todos: ITodo[]) {
		this.isLoading = false;
		this.originalList = todos; 
		this.displayList = todos;
		this.displayMode = DisplayMode.All;
		this.effectTrigger = null;
		this.search = {
			searchTerm: '',
		};
		this.filter = {
			state: StateFilter.all,
		} as IFilter;
		this.sort = {
			column: 'id',
			direction: SortDirection.Asc
		} as ISort;
		this.paging = {
			totalCount: todos.length,
			activePage: todos.length > 0 ? 1 : 0,
			startIndex: 0,
			endIndex: todos.length > 5 ? 5 : todos.length,
			itemsPerPage: 5
		} as IPaging;
		this.settings = {
			general: {
				isConfirmEnabled: true,
				isPaginationEnabled: true,
				isInfiniteScrollEnabled: false,
				listSizeType: ListContainerType.Dynamic,
				fixedListSize: 200
			} as IGeneralSettings,
			search: {
        isSearchOnKeyPressEnabled: false,
				debounceTime: 500
			} as ISearchSettings,
			pagination: {
				paginationType: PaginationType.Classic,
				maxVisiblePages: 3
			} as IPaginationSettings
		} as ISettings;
		this.activeTab = 'add-todo';
	}
}
