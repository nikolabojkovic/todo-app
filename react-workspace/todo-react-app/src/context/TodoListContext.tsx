import { createContext, useContext, useReducer } from "react";

import { IAction, TodoActions, IPaging, ITodo, StateFilter, ISettings } from "../models";
import { IState } from "./";
 
export const TodosContext = createContext({} as IState);
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const TodosDispatchContext = createContext(null as any);

type Props = {
  children: JSX.Element | JSX.Element[],
  initialState: IState
}

export function TodoStateProvider({ initialState, children }: Props) {
  const [state, dispatch] = useReducer(
    todoStateReducer,
    initialState
  );

  return (
    <TodosContext.Provider value={state}>
      <TodosDispatchContext.Provider value={dispatch}>
        {children}
      </TodosDispatchContext.Provider>
    </TodosContext.Provider>
  );
}

export function useTodoList() {
  return useContext(TodosContext);
}

export function useTodoListDispatch() {
  return useContext(TodosDispatchContext);
}

export default TodosDispatchContext.Provider;

export function todoStateReducer(state: IState, action: IAction) {
  switch (action.type) { 
    case TodoActions.loadingStarted: {
      return { 
        ...state,
        effectTrigger: null,
        isLoading: true
      } as IState;
    }
    case TodoActions.activeTabChanged: {
      return { 
        ...state,
        activeTab: action.payload.activeTab
      } as IState;
    }
    case TodoActions.fetch: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.fetch, payload: action.payload },
      } as IState;
    }
    case TodoActions.filter: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.filter, payload: action.payload },
      } as IState;
    }
    case TodoActions.search: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.search, payload: action.payload },
      } as IState;
    }
    case TodoActions.sort: {
      return { 
        ...state,
        effectTrigger: { type: TodoActions.sort, payload: action.payload },
      } as IState;
    }

    case TodoActions.fetched: {
      return {
        ...state,
        isLoading: false,
        originalList: action.payload.list,
        displayList: action.payload.list,
        effectTrigger: null,
        paging: {
          ...state.paging,
          totalCount: action.payload.list.length,
          activePage: state.paging.activePage === 0 ? 1 : state.paging.activePage,
          itemsPerPage: state.paging.itemsPerPage,
          startIndex: ((state.paging.activePage === 0 ? 1 : state.paging.activePage) - 1) * state.paging.itemsPerPage,
          endIndex: (state.paging.activePage === 0 ? 1 : state.paging.activePage) * state.paging.itemsPerPage
        } as IPaging,
      };
    }
    case TodoActions.searched: {
      return {
        ...state,
        isLoading: false,
        effectTrigger: null,
        displayList: [...action.payload.list],
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      };
    }
    case TodoActions.filtered: {
      return {
        ...state,
        isLoading: false,
        effectTrigger: null,
        displayList: [...action.payload.list],
        filter: {...action.payload.filter},
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      };
    }
    case TodoActions.sorted: {
      return {
        ...state,
        isLoading: false,
        effectTrigger: null,
        displayList: [...action.payload.list],
        sort: {...action.payload.sort},
        paging: {...state.paging} as IPaging
      };
    }
    case TodoActions.imported: {
      return {
        ...state,
        originalList: [...action.payload.list],
        displayList: [...action.payload.list],
        effectTrigger: { type: TodoActions.imported },
        search: { searchTerm: '' },
        filter: {
          state: StateFilter.all
        },
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          totalCount: action.payload.list.length,
          startIndex: (action.payload.activePage - 1) * state.paging.itemsPerPage,
          endIndex: action.payload.activePage * state.paging.itemsPerPage
        } as IPaging
      };
    }

    case TodoActions.searchTermUpdated: {
      return {
        ...state,
        effectTrigger: null,
        search: { searchTerm: action.payload.searchTerm },
      };
    }    
    case TodoActions.pagingFatched: {
      return {
        ...state,
        effectTrigger: null,
        paging: { 
          ...state.paging,
          totalCount: action.payload.totalCount,
          activePage: action.payload.activePage,
          itemsPerPage: action.payload.itemsPerPage,
          startIndex: (action.payload.activePage - 1) * action.payload.itemsPerPage,
          endIndex: action.payload.activePage * action.payload.itemsPerPage
        } as IPaging
      } as IState;
    }
    case TodoActions.pagingUpdated: {
      return {
        ...state,
        effectTrigger: { type: TodoActions.pagingUpdated },
        paging: {
          ...state.paging,
          activePage: action.payload.activePage,
          itemsPerPage: action.payload.itemsPerPage,
          startIndex: (action.payload.activePage - 1) * action.payload.itemsPerPage,
          endIndex: action.payload.activePage * action.payload.itemsPerPage
        } as IPaging
      } as IState;
    }
    case TodoActions.settingsFetched: {
      return {
        ...state,
        effectTrigger: null,
        settings: {
          ...action.payload
        } as ISettings
      };
    }
    case TodoActions.settingsUpdated: {
      return {
        ...state,
        effectTrigger: { type: TodoActions.settingsUpdated },
        settings: {
          ...action.payload
        } as ISettings
      };
    }

    case TodoActions.added: {
      const id = state.originalList.length >= 1 
      ? state.originalList
          .sort((a: ITodo, b: ITodo) => a.id > b.id ? 1 : -1)[state.originalList.length - 1].id 
      : 0;
      const newTodo = {
        id: id + 1,
        title: action.payload.title,
        description: action.payload.description,
        completed: false, 
        createdAt: action.payload.createdAt
      } as ITodo;

      return {
        ...state,
        originalList: [...state.originalList, newTodo],
        displayList: [...state.displayList, newTodo],
        effectTrigger: { type: TodoActions.added },
        paging: {
          ...state.paging, 
          totalCount: state.paging.totalCount + 1,
          activePage: calculateActivePageOnAdd(state.paging),
          startIndex: (calculateActivePageOnAdd(state.paging) - 1) * state.paging.itemsPerPage,
          endIndex: calculateActivePageOnAdd(state.paging) * state.paging.itemsPerPage
        } as IPaging
      } as IState;
    }
    case TodoActions.changed: {
      return {
        ...state,
        originalList: state.originalList.map(t => {
          if (t.id === action.payload.todo.id) {
            return action.payload.todo;
          } else {
            return t;
          }
        }),
        displayList: state.displayList.map(t => {
          if (t.id === action.payload.todo.id) {
            return action.payload.todo;
          } else {
            return t;
          }
        }),
        effectTrigger: { type: TodoActions.changed },
        paging: {...state.paging}
      } as IState;
    }
    case TodoActions.deleted: {
      return {
        ...state,
        originalList: state.originalList.filter(t => t.id !== action.payload.id),
        displayList: state.displayList.filter(t => t.id !== action.payload.id),
        effectTrigger: { type: TodoActions.deleted },
        paging: {
          ...state.paging, 
          totalCount: state.paging.totalCount - 1,
          activePage: calculateActivePageOnDelete(state.paging),
          startIndex: (calculateActivePageOnDelete(state.paging) - 1) * state.paging.itemsPerPage,
          endIndex: calculateActivePageOnDelete(state.paging) * state.paging.itemsPerPage,  
        } as IPaging
      } as IState;
    }
    default: {
      throw Error('Unknown action: ' + action.type);
    }
  }
}

function calculateActivePageOnDelete(paging: IPaging) {
  return Math.ceil((paging.totalCount - 1) / paging.itemsPerPage) < paging.activePage
    ? paging.activePage - 1 : paging.activePage;
}

function calculateActivePageOnAdd(paging: IPaging) {
  return Math.ceil((paging.totalCount + 1) / paging.itemsPerPage) > paging.activePage
    ? Math.ceil((paging.totalCount + 1) / paging.itemsPerPage) : paging.activePage;
}