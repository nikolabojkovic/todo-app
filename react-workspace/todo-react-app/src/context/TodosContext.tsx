import { createContext, useContext, useReducer } from "react";
import { IPaging } from "../models/IPaging";
import { ITodoList } from "../models/ITodoList";
import { ITodo } from "../models/Todo";
import { TodoService } from "../services/TodoService";
 
export const TodosContext = createContext({} as ITodoList);
export const TodosDispatchContext = createContext(null as any);
export const todoService: TodoService = new TodoService();

export function TodoListProvider({ children }: any) {
  const [todoList, dispatch] = useReducer(
    todoListReducer,
    todoService.getTodoList()
  );

  return (
    <TodosContext.Provider value={todoList}>
      <TodosDispatchContext.Provider value={dispatch}>
        {children}
      </TodosDispatchContext.Provider>
    </TodosContext.Provider>
  );
}

export function useTodoList() {
  return useContext(TodosContext);
}

export function useTodoListDispatch() {
  return useContext(TodosDispatchContext);
}

function todoListReducer(todoList: ITodoList, action: any) {
  switch (action.type) {
    case 'added': {
      const id = todoList.originalList.length >= 1 
      ? todoList.originalList
          .sort((a: ITodo, b: ITodo) => a.id > b.id ? 1 : -1)[todoList.originalList.length - 1].id 
      : 0;
      const newTodo = {
        id: id + 1,
        title: action.title,
        description: action.description,
        completed: false, 
        createdAt: new Date()
      } as ITodo;

      return {
        ...todoList,
        originalList: [...todoList.originalList, newTodo],
        displayList: [...todoList.displayList, newTodo],
        paging: {
          ...todoList.paging, 
          totalCount: todoList.paging.totalCount + 1,
          activePage: calculateActivePageOnAdd(todoList.paging),
          startIndex: (calculateActivePageOnAdd(todoList.paging) - 1) * todoList.paging.itemsPerPage,
          endIndex: calculateActivePageOnAdd(todoList.paging) * todoList.paging.itemsPerPage
        } as IPaging
      } as ITodoList;
    }
    case 'changed': {
      return {
        ...todoList,
        originalList: todoList.originalList.map(t => {
          if (t.id === action.todo.id) {
            return action.todo;
          } else {
            return t;
          }
        }),
        displayList: todoList.displayList.map(t => {
          if (t.id === action.todo.id) {
            return action.todo;
          } else {
            return t;
          }
        }),
        paging: {...todoList.paging}
      } as ITodoList;
    }
    case 'deleted': {
      return {
        ...todoList,
        originalList: todoList.originalList.filter(t => t.id !== action.id),
        displayList: todoList.displayList.filter(t => t.id !== action.id),
        paging: {
          ...todoList.paging, 
          totalCount: todoList.paging.totalCount - 1,
          activePage: calculateActivePageOnDelete(todoList.paging),
          startIndex: (calculateActivePageOnDelete(todoList.paging) - 1) * todoList.paging.itemsPerPage,
          endIndex: calculateActivePageOnDelete(todoList.paging) * todoList.paging.itemsPerPage,  
        } as IPaging
      } as ITodoList;
    }
    case 'paging-updated': {
      return {
        ...todoList,
        paging: {
          ...todoList.paging,
          activePage: action.activePage,
          itemsPerPage: action.itemsPerPage,
          startIndex: (action.activePage - 1) * action.itemsPerPage,
          endIndex: action.activePage * action.itemsPerPage
        } as IPaging
      } as ITodoList
    }
    case 'searched': {
      const filteredList = todoService.filter(todoList.originalList, todoList.filter);
      const searchedList = todoService.search(filteredList, action.searchTerm);
      return {
        ...todoList,
        displayList: [...searchedList],
        search: { searchTerm: action.searchTerm },
        paging: {
          ...todoList.paging,
          activePage: action.activePage,
          totalCount: searchedList.length,
          startIndex: (action.activePage - 1) * todoList.paging.itemsPerPage,
          endIndex: action.activePage * todoList.paging.itemsPerPage
        } as IPaging
      }
    }
    case 'searchTerm-updated': {
      return {
        ...todoList,
        search: { searchTerm: action.searchTerm },
      }
    }
    case 'filtered': {
      const filteredList = todoService.filter(todoList.originalList, action.filter);
      const searchedList = todoService.search(filteredList, todoList.search.searchTerm);
      return {
        ...todoList,
        displayList: [...searchedList],
        filter: {...action.filter},
        paging: {
          ...todoList.paging,
          activePage: action.activePage,
          totalCount: searchedList.length,
          startIndex: (action.activePage - 1) * todoList.paging.itemsPerPage,
          endIndex: action.activePage * todoList.paging.itemsPerPage
        } as IPaging
      }
    }
    case 'sorted': {
      const filteredList = todoService.filter(todoList.originalList, todoList.filter);
      const searchedList = todoService.search(filteredList, todoList.search.searchTerm);
      const sortedList = todoService.sort(searchedList, action.sort);

      return {
        ...todoList,
        displayList: [...sortedList],
        sort: {...action.sort},
        paging: {...todoList.paging} as IPaging
      }
    }
    case 'imported': {
      return {
        ...todoList,
        originalList: [...action.originalList],
        displayList: [...action.originalList],
        search: { searchTerm: '' },
        filter: {
          completed: false,
          uncompleted: false
         },
        paging: {
          ...todoList.paging,
          activePage: action.activePage,
          totalCount: action.originalList.length,
          startIndex: (action.activePage - 1) * todoList.paging.itemsPerPage,
          endIndex: action.activePage * todoList.paging.itemsPerPage
        } as IPaging
      }
    }
    default: {
      throw Error('Unknown action: ' + action.type);
    }
  }
}

function calculateActivePageOnDelete(paging: IPaging) {
  return Math.ceil((paging.totalCount - 1) / paging.itemsPerPage) < paging.activePage
    ? paging.activePage - 1 : paging.activePage;
}

function calculateActivePageOnAdd(paging: IPaging) {
  return Math.ceil((paging.totalCount + 1) / paging.itemsPerPage) > paging.activePage
    ? Math.ceil((paging.totalCount + 1) / paging.itemsPerPage) : paging.activePage;
}