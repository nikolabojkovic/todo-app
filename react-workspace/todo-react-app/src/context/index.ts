export * from './IState';
export * from './TodoListContext';
export * from './testData';