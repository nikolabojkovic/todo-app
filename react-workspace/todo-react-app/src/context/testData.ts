import { ITodo } from "../models/Todo";
import { IPaging } from "../models/IPaging";
import { IState } from "./IState";
import { ISort, SortDirection } from "../models/ISort";
import { IFilter, StateFilter } from "../models/IFilter";
import { IGeneralSettings, IPaginationSettings, ISearchSettings, ISettings, ListContainerType, PaginationType } from "../models/ISettings";

export const todoListTestData: ITodo[] = [
  {
    id: 1,
    title: "Task 1",
    description: "Description 1",
    completed: false,
    createdAt: new Date(2022, 1, 4)
  } as ITodo,
  {
    id: 2,
    title: "Task 2",
    description: "Description 2",
    completed: true,
    createdAt: new Date(2022, 1, 5)
  } as ITodo,
  {
    id: 3,
    title: "Task 3",
    description: "Description 3",
    completed: false,
    createdAt: new Date(2022, 1, 6)
  } as ITodo,
  {
    id: 4,
    title: "Task 4",
    description: "Description 4",
    completed: false,
    createdAt: new Date(2022, 1, 7)
  } as ITodo,
  {
    id: 6,
    title: "Task 6",
    description: "Description 6",
    completed: false,
    createdAt: new Date(2022, 2, 4)
  } as ITodo,
  {
    id: 5,
    title: "Task 5",
    description: "Description 5",
    completed: false,
    createdAt: new Date(2022, 3, 8)
  } as ITodo
] as ITodo[];

export const stateTestData: IState = 
{
  originalList: todoListTestData,
  displayList: todoListTestData,
  search: { searchTerm: '' },
  filter: { state: StateFilter.all } as IFilter,
  sort: { column: 'createdAt', direction: SortDirection.Asc} as ISort,
  paging: {
    totalCount: 6,
    activePage: 1,
    itemsPerPage: 5,
    startIndex: 0,
    endIndex: 5,
  } as IPaging,
  settings: {
    general: {
      isConfirmEnabled: true,
      isPaginationEnabled: true,
      isInfiniteScrollEnabled: false,
      listSizeType: ListContainerType.Dynamic,
      fixedListSize: 200
    } as IGeneralSettings,
    search: {
      isSearchOnKeyPressEnabled: false,
      debounceTime: 500
    } as ISearchSettings,
    pagination: {
      paginationType: PaginationType.Classic,
      maxVisiblePages: 3
    } as IPaginationSettings
  } as ISettings
} as IState;