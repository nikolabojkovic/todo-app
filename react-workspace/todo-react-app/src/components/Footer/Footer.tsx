import { useRef } from 'react';
import { Container, Row } from 'react-bootstrap';
import { useTodoList } from '../../context/TodoListContext';
import { PageSize } from '../Paging/PageSize';
import { Pagination } from '../Paging/Pagination';

export function Footer() {
  const todoList = useTodoList();

  const inputSelectRef = useRef<HTMLSelectElement | null>(null);
  const pageCount = Math.ceil(todoList.paging.totalCount / todoList.paging.itemsPerPage);
  const rotate = false;
  const maxVisiblePagesCount = 3

  return (
    <section className="footer p-0 mt-2">      
      <Container fluid>
        <Row xs={1} sm={3}>
          <PageSize 
            pageCount={pageCount} 
            inputSelectRef={inputSelectRef}
          />
          <Pagination 
            pageCount={pageCount} 
            inputSelectRef={inputSelectRef} 
            maxVisiblePagesCount={maxVisiblePagesCount} 
            rotate={rotate} 
          />
        </Row>
      </Container>
    </section>
  )
}