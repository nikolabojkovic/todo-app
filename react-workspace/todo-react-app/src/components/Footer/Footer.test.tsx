import renderer from 'react-test-renderer';
import { Footer } from './Footer';
import { TodoStateProvider } from '../../context/TodoListContext';
import { stateTestData } from '../../context/testData';

describe('Paging', () => {
  it('component should match snapshot', () => {
    const jsxElement = 
    (<TodoStateProvider todoList={stateTestData}>
      <Footer />
    </TodoStateProvider>);
    const tree = renderer.create(jsxElement).toJSON();

    expect(tree).toMatchSnapshot();
  });
});