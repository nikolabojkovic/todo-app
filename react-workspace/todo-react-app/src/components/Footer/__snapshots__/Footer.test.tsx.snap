// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paging component should match snapshot 1`] = `
<section
  className="footer p-0 mt-2"
>
  <div
    className="container-fluid"
  >
    <div
      className="row row-cols-sm-3 row-cols-1"
    >
      <div
        className="p-3 pt-2 pb-2 d-flex justify-content-sm-start justify-content-center align-items-center col-sm-3"
      >
        Page 
        1
         of 
        2
      </div>
      <div
        className="p-3 pt-2 pb-2 col-sm-4"
      >
        <div
          className="justify-content-center hstack gap-2"
        >
          <div
            className="ms-sm-auto"
          >
            Page size:
             
          </div>
          <div
            className=""
          >
            <div
              className="dropdown"
            >
              <button
                aria-expanded={false}
                className="action-button dropdown-toggle btn btn-outline-secondary"
                data-testid="selected-size-option"
                disabled={false}
                id="dropdown-basic"
                onClick={[Function]}
                type="button"
              >
                 
                5
                 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="p-3 pt-2 pb-2 d-flex justify-content-sm-end justify-content-center col-sm-5"
      >
        <ul
          className="d-flex align-items-center pagination pagination-sm"
        >
          <li
            className="page-item disabled"
          >
            <span
              className="page-link"
              data-testid="first-page"
              onClick={[Function]}
            >
              <span
                aria-hidden="true"
              >
                «
              </span>
              <span
                className="visually-hidden"
              >
                First
              </span>
            </span>
          </li>
          <li
            className="page-item disabled"
          >
            <span
              className="page-link"
              data-testid="prev-page"
              onClick={[Function]}
            >
              <span
                aria-hidden="true"
              >
                ‹
              </span>
              <span
                className="visually-hidden"
              >
                Previous
              </span>
            </span>
          </li>
          <li
            className="page-item active"
          >
            <span
              className="page-link"
              data-testid="page-1"
              onClick={[Function]}
            >
              1
              <span
                className="visually-hidden"
              >
                (current)
              </span>
            </span>
          </li>
          <li
            className="page-item"
          >
            <a
              className="page-link"
              data-testid="page-2"
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex={0}
            >
              2
            </a>
          </li>
          <li
            className="page-item"
          >
            <a
              className="page-link"
              data-testid="next-page"
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                aria-hidden="true"
              >
                ›
              </span>
              <span
                className="visually-hidden"
              >
                Next
              </span>
            </a>
          </li>
          <li
            className="page-item"
          >
            <a
              className="page-link"
              data-testid="last-page"
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                aria-hidden="true"
              >
                »
              </span>
              <span
                className="visually-hidden"
              >
                Last
              </span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</section>
`;
